packages/openapi-generator/src/options/options-per-service.ts

```diff
@@ -66,76 +65,120 @@ export async function getOptionsPerService(
+      parseDirectoryName(relativePath);
+
+    directoryNamesByPaths[inputPath] = directoryName;
+    return directoryNamesByPaths;
+  }, {});
+}
+
+function getPathsByDirName(
+  dirNamesByPaths: Record<string, string>
+): Record<string, string[]> {
+  return Object.entries(dirNamesByPaths).reduce(
+    (pathsByDirName, [inputPath, dirName]) => {
+      if (!pathsByDirName[dirName]) {
+        pathsByDirName[dirName] = [];
+      }
+      pathsByDirName[dirName].push(inputPath);
+      return pathsByDirName;
+    },
+    {}
+  );
+}
+
+function validateDirectoryNames(dirNamesByPaths: Record<string, string>): void {
+  const originalDirNames = Object.values(dirNamesByPaths);
+  const uniqueDirNames = unique(originalDirNames);
+  const hasDuplicates = originalDirNames.length !== uniqueDirNames.length;
+
+  if (hasDuplicates) {
+    const pathsByDirName = getPathsByDirName(dirNamesByPaths);
+    const duplicates = Object.entries(pathsByDirName).filter(
+      ([, paths]) => paths.length > 1
     );
-  }
 
-  return optionsPerService;
+    const duplicatesList = duplicates
+      .map(
+        ([dirName, paths]) =>
+          `\t\tDirectory name: '${dirName}', specifications: [\n${paths
+            .map(path => `\t\t\t${path}`)
+            .join(',\n')}\n\t\t]`
+      )
+      .join('\n');
+
+    const errorMessage = `Duplicate service directory names found. Customize directory names with \`optionsPerService\` or enable automatic name adjustment with \`skipValidation\`.\n\tDuplicates:\n${duplicatesList}`;
+    throw new Error(errorMessage);
+  }
 }
 
 /**
- * Parse the name of the service based on the file path.
+ * Parse the name of the service directory based on the file path.
  * @param filePath Path of the service specification.
  * @returns The parsed name.
  */
-function parseServiceName(filePath: string): string {
+function parseDirectoryName(filePath: string): string {
   return parse(filePath).name.replace(/-openapi$/, '');
 }
 
 /**
  * Get the options for one service based on the options per service and the input file path.
  * If the file path does not exist in the options a default config is created.
  * If the service options for a file path are given only partially, default values are added for the missing values.
- * @param optionsPerService The original options per service to get the service options from.
- * @param relativeInputFilePath The input file path for which to find service options.
- * @param serviceName The default name of the according service.
+ * @param serviceOptions The original options for this service as specified in the per service options.
+ * @param directoryName The directory name of the according service.
  * @returns Service options.
  */
 export function getServiceOptions(
-  optionsPerService: PartialOptionsPerService,
-  relativeInputFilePath: string,
-  serviceName: string
+  serviceOptions: Partial<ServiceOptions> | undefined,
+  directoryName: string
```

Comment:
```suggestion
  directoryName: string,
  serviceOptions?: Partial<ServiceOptions>
```

Would changing the order to allow `serviceOptions` to be optional instead of `... | undefined` be worth it?

Reply:
Not necessarily, but it looks slightly nicer 😊